.github/docs/source/config.rst
Configuration
=============

Environment Setup
-----------------
  
Here's an Example of ``sample.env``

.. code-block:: bash
  
   API_ID = 
   API_HASH = 
   BOT_TOKEN = 
   MONGO_DB_URI = 
   LOG_GROUP_ID = 
   OWNER_ID = 
   STRING_SESSIONS = 
  
- Copy the ``example.env`` file to ``.env``
   .. code-block:: bash

      cp sample.env .env

- Fill in the necessary variables in your ``.env`` :
   - **API_ID**: The API ID obtained from `Telegram <https://my.telegram.org/auth>`_.
   - **API_HASH**: The API hash obtained from `Telegram <https://my.telegram.org/auth>`_.
   - **BOT_TOKEN**: The bot token you get from `BotFather <https://core.telegram.org/bots#botfather>`_.
   - **MONGO_DB_URI**: The MongoDB connection string for your database.
   - **LOG_GROUP_ID**: The ID of the group to which logs should be sent (set to `0` if no logs are needed).
   - **OWNER_ID**: The  User ID That will treated as the Owner of the bot, Multiple Id can be separated with a space.
   - **STRING_SESSIONS**: Your session string (usually generated via a method like Pyrogram) or You can generate from `Telegram Tools <https://telegram.tools/session-string-generator#pyrogram>`_ And make sure you environment is **Production** don't use **Test**.

You also can fill these :ref:`Optinal  Vars<Optiona_Variables_recommended>`

Available Vars
--------------

Here is a List of all Available Vars of `YukkiMusic <https://github.com/TheTeamVivek/YukkiMusic>`_.

Config vars are basically the variables which configure or modify bot to function, which are the basic necessities of plugins or code to work. You have to set the proper Necessary vars to make it functional and to start the basic feature of bot.

Necessary Vars
^^^^^^^^^^^^^^

- API_ID & API_HASH
   - Go to `my.telegram.org <https://my.telegram.org/auth>`_ then Enter your Phone Number with your country code.

   - After, you are logged in click on API Development Tools.

   - Enter Anything as App name and App short name, Enter `my.telegram.org` in url section

   - That’s it, You'll get your API_ID and API_HASH.

- BOT_TOKEN
   Get token of the bot from the `@Botfather <https://t.me/Botfather>`_ in Telegram

- LOG_GROUP_ID
   .. important::

      You'll need a **Group** for this. 

      Remember to add your Music Bot , Assistant Accounts and Logger Id in Group and Promote them Admin with Full Rights.
   - Add `@MissRose_Bot <https://t.me/MissRose_Bot>`_ in your Group from Add Member > Search ``@MissRose_Bot`` and then Add.

   - After added, Just type ``/id`` in the chat.

   - You'll get the ID of your group.

- OWNER_ID
   .. note::

      Value must be an integer like 0123456789

      You can add multiple userid seperated with a Space

      .. code-block:: bash
         :caption: Example of Multiple Userid

          OWNER_ID = 1234567890 2345617890 8790654321 6578012347

   Your user id (not username) Get it by using command /id on the Group in the reply to your message where Rose Bot was added.

- STRING_SESSIONS
   A list of Pyrogram String Session seperated with comma "," of a Telegram Account which will be joining Group Calls for streaming.

   Your session string (usually generated via a method like Pyrogram) or You can generate from `Telegram Tools <https://telegram.tools/session-string-generator#pyrogram>`_ And make sure you environment is **Production** donn't use **Test**.

   .. code-block:: bash
         :caption: Example of Multiple String Sessions

          STRING_SESSIONS = string1,  string2, string3,  string4

   Like this as your mood you can add multiple String sessions of Your assistant for multiple Assistsant.

.. _Optiona_Variables_recommended:

Optional Variables But  Recommended
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

- MONGO_DB_URI
       Not a Necessary var, but yes kind off.
   .. note::

      Yukki no longer requires MONGO DB as mandatory. Leave it blank and bot will use Yukki’s database for your bot. Seperate database and Easy to use.

      To maintain bot’s privacy you wont be able to manage sudoers.  Bot will create an separate collection for you and no other bot's database will clash with it.

- COOKIE_LINK
   .. important::

      This is **not a Necessary** variable, but it is necessary for the bot to play songs perfectly due to YouTube verification.  

      Without cookies, the bot may be unable to download songs.  

   **How to obtain COOKIE_LINK:**  

   1. Get your YouTube cookies.  

   2. Go to `batbin.me <https://batbin.me>`_.

   3. Paste your cookies and tap on **Save**.  

   4. Copy the generated URL.  

   5. Set ``COOKIE_LINK`` with this URL.  

   **Alternative Method:**  

   If you don't want to use a URL, you can manually add cookies:  

   1. Navigate to `config/cookies/ <https://github.com/TheTeamVivek/YukkiMusic/tree/dev/config/cookies>`_.  

   2. Create a ``.txt`` file.  

   3. Paste your cookies inside the file.  

   This ensures smooth song playback without verification issues.  

   .. seealso::
  
      Don't know how to get cookies? See :doc:`faq/cookies` This properly


Optional Variables
^^^^^^^^^^^^^^^^^^

.. _auto_leaving_assistant:

- AUTO_LEAVING_ASSISTANT
   .. admonition:: What is This?
      :class: hint

      Yukki has a feature called auto assistant leaving.

      It leaves assistant accounts from all chats after a certain amount of time. You can set that time via  :ref:`ASSISTANT_LEAVE_TIME<assistant_leave_time>`

      You can use this mode, it enhances your bot's performance 

   Set it in **True** if you want to leave your assistant after a certain amount of time.. Default to **False**.

.. _assistant_leave_time:

- ASSISTANT_LEAVE_TIME
   Time after which your assistant account will leave served chats automatically. Default to ``1800`` seconds, i.e 30 Mins

   .. note::

      To Use this feature, use :ref:`AUTO_LEAVING_ASSISTANT<auto_leaving_assistant>`. Set it to ``True`` to enable automatic leaving. When set to ``False``, the assistant will remain in served chats indefinitely.

- CLEANMODE_MINS
   .. admonition:: What is This?
      :class: hint

      Cleanmode, that helps Bot can deletes its old messages automatically  in the group after a set period. You can configure the duration here.

   Time (in minutes) after which bot will delete its old messages from chats. **Default to 5 Mins.**

- DURATION_LIMIT_MIN
   .. admonition:: What is This?
      :class: hint

      When a user plays any music from youtube , apple , spotify or any platform , it gets downloaded on the bot.  You can restrict users of your bot to play upto a maximum limit of audio duration. 

      For example : If ``DURATION_LIMIT`` is 60 mins, so if a user tries to play a movie of 2-3 hours you're bot won't play that.

   Custom max audio(music) duration for voice chat. Default to 60 mins.

- EXTRA_PLUGINS
   .. admonition:: What is This?
      :class: hint

      This variable controls whether extra plugins should be loaded by the bot.

   Set ``True`` to enable the extra plugin s, (default: ``False``).


CHANGELOG.md
# CHANGELOG

All notable changes to YukkiMusic will be documented in this file.


## [v1.3] - 2024-11-15

### Added
- Added the `asyncify` decorator to convert synchronous functions to asynchronous.
- Added the `OWNER` command that will show in the log group or bot PM/DM.

### Removed
- Deprecated the `oauth2` method, which is no longer supported. This functionality has been merged into `yt-dlp`.

### Fixed
- Various bug fixes and performance improvements.

**Full Changelog:** [`v1.2...v1.3`](https://github.com/TheTeamVivek/YukkiMusic/compare/v1.2...v1.3)

## [v1.2] - 2024-11-03

### Added
- Added Multiple Languages Support for commands
- Multiple language support for bot Help Menu [ Only for primary plugins Not for External Plugins ]
- All commands can be used without prefix [ Except English commands ]
- User can Request her data and can Delete [ Except: Chat, Banned Users, Blacklist Chats]
- PRIVACY.md For YukkiMusic
### Changed
- `Apple`, `Carbon`, `Saavn`, `Resso`, `SoundCloud`, `Spotify`, `Telegram`, `YouTube` are centralized to a class [PlaTForms](https://github.com/TheTeamVivek/YukkiMusic/blob/master/YukkiMusic%2Fplatforms%2F__init__.py)
- Explained Privacy policy in `/privacy` command
- Now Assistsant will joinchat when chat is private
- Now User Friendly README.md

**Full Changelog:** [`v1.1...v1.2`](https://github.com/TheTeamVivek/YukkiMusic/compare/v1.1...v1.2)

## [v1.1] - 2024-10-14

### Added
- Unlimited assistant support for handling multiple voice chats
- Mongodb Data Export/import Support 
- Added JioSaavn Playback support 
- Added yt-dlp-youtube-oauth2 to bypass Singin Issue
- The currently playing message will be deleted when switching to the next track.

### Changed
- Updated Python Version to 3.12.7-slim
- Improved error handling in music playback
- Enhanced queue management system
- Better formatting for duration display
- Optimized database operations

### Fixed
- Delete Files after streams end
- Updated `langs/en.yml` Standardized to use English letters instead of mini caps.
- Commands are now sourced from `command.yml` Any updates to plugin commands will automatically update the help message

### Removed

- Some unused plugins vars.py, groupass.py, player.py.
-  Assets folder due to lack of use.
- Unused dependencies from requirements.txt

**Full Changelog:** [`v1.0...v1.1`](https://github.com/TheTeamVivek/YukkiMusic/compare/v1.0...v1.1)

## [v1.0] - 2024-10-05


- Initial release of YukkiMusic
- Thanks To [Pranav-Saraswat](https://github.com/Pranav-Saraswat) For Their YukkiMusicFork For Making Working 
- Thanks To [TeamYukki](https://github.com/TeamYukki/) for Their [YukkiMusicBot](https://github.com/TeamYukki/YukkiMusicBot)

### Features
- High quality music streaming
- Video streaming capability
- Interactive inline buttons
- Detailed playback statistics
- Group management commands
- Customizable bot settings

### Notes
- Base version established with core functionality
- Compatible with Python 3.9+
- Built with Pyrogram and py-tgcalls

config/config.py
#
# Copyright (C) 2024-2025 by TheTeamVivek@Github, < https://github.com/TheTeamVivek >.
#
# This file is part of < https://github.com/TheTeamVivek/YukkiMusic > project,
# and is released under the MIT License.
# Please see < https://github.com/TheTeamVivek/YukkiMusic/blob/master/LICENSE >
#
# All rights reserved.

import re
import sys
from os import getenv

from dotenv import load_dotenv
from pyrogram import filters

load_dotenv()

TRUE_VALUES = ("true", "1", "yes")


def is_true(value: str) -> bool:
    return value.lower() in TRUE_VALUES


# To know what each variable does,
# check out https://theteamvivek.github.io/YukkiMusic/config.html).

# -------------------
# Necessary Variables
# -------------------


# Get it from my.telegram.org

API_ID = int(getenv("API_ID", "0"))

API_HASH = getenv("API_HASH")

## Get it from @Botfather in Telegram.
BOT_TOKEN = getenv("BOT_TOKEN")

# You'll need a Private Group ID for this.
LOG_GROUP_ID = getenv("LOG_GROUP_ID", "")

# Your User ID.
OWNER_ID = list(
    map(int, getenv("OWNER_ID", "6815918609").split())
)  # Input type must be interger, Replace 6815918609 it to your own id

# You'll need a Pyrogram String Session for these vars. Generate String from telegram.tools

raw_sessions = getenv("STRING_SESSIONS")

# Split the sessions only if raw_sessions is not empty
STRING_SESSIONS = list(map(str.strip, raw_sessions.split(","))) if raw_sessions else []

# ------------------------------------
# Optional But Needed [ You Can Skip ]
# ------------------------------------

# Your cookies pasted link on batbin.me
# you can skip if you are adding cookies
# manually in config/cookies dir
COOKIE_LINK = getenv("COOKIE_LINK", None)

# Database to save your chats and stats...
# Get MongoDB:-  https://telegra.ph/How-To-get-Mongodb-URI-04-06
MONGO_DB_URI = getenv("MONGO_DB_URI", None)


# -------------- |
# Fully Optional |
# -------------- |


# Set it in True if you want to leave your assistant after
# a certain amount of time. [Set time via AUTO_LEAVE_ASSISTANT_TIME]
AUTO_LEAVING_ASSISTANT = is_true(getenv("AUTO_LEAVING_ASSISTANT", "False"))

# Time after which you're assistant account
# will leave chats automatically.
AUTO_LEAVE_ASSISTANT_TIME = int(
    getenv("ASSISTANT_LEAVE_TIME", "1800")
)  # Remember to give value in Seconds

CLEANMODE_DELETE_MINS = int(
    getenv("CLEANMODE_MINS", "5")  #  TODO MAKE the env or local variable name same
)  # Remember to give value in Seconds

# Custom max audio(music) duration for voice chat.
# set DURATION_LIMIT in variables with your own time(mins),
# Default to 60 mins.

DURATION_LIMIT_MIN = int(
    getenv("DURATION_LIMIT", "60")
)  # Remember to give value in Minutes

EXTRA_PLUGINS = is_true(getenv("EXTRA_PLUGINS", "False"))

# Fill False if you Don't want to load extra plugins


EXTRA_PLUGINS_REPO = getenv(
    "EXTRA_PLUGINS_REPO",
    "https://github.com/TheTeamVivek/Extra-Plugin",
)
# Fill here the external plugins repo where plugins that you want to load

# Your Github Repo.. Will be shown on /start Command
GITHUB_REPO = getenv("GITHUB_REPO", None)

# GIT TOKEN ( if your edited repo is private
GIT_TOKEN = getenv(
    "GIT_TOKEN",
    "",
)

# Get it from http://dashboard.heroku.com/account
HEROKU_API_KEY = getenv("HEROKU_API_KEY")

# You have to Enter the app name
# which you gave to identify your  Music Bot in Heroku.
HEROKU_APP_NAME = getenv("HEROKU_APP_NAME")

# MaximuM limit for fetching playlist's track from youtube, spotify, apple links.
PLAYLIST_FETCH_LIMIT = int(getenv("PLAYLIST_FETCH_LIMIT", "25"))

# Set it true if you want your bot to be private only
# [You'll need to allow CHAT_ID via /authorize command then only your bot will play music in that chat.]
PRIVATE_BOT_MODE = is_true(getenv("PRIVATE_BOT_MODE", "False"))

# If you want your bot to setup the commands automatically in the bot's menu set it to true.
# Refer to https://i.postimg.cc/Bbg3LQTG/image.png
SET_CMDS = is_true(getenv("SET_CMDS", "False"))

# Maximum Limit Allowed for users to save playlists on bot's server
SERVER_PLAYLIST_LIMIT = int(getenv("SERVER_PLAYLIST_LIMIT", "25"))

# Duration Limit for downloading Songs in MP3 or MP4 format from bot
SONG_DOWNLOAD_DURATION = int(
    getenv("SONG_DOWNLOAD_DURATION_LIMIT", "90")
)  # Remember to give value in Minutes

# Spotify Client.. Get it from https://developer.spotify.com/dashboard
SPOTIFY_CLIENT_ID = getenv("SPOTIFY_CLIENT_ID", "19609edb1b9f4ed7be0c8c1342039362")
SPOTIFY_CLIENT_SECRET = getenv(
    "SPOTIFY_CLIENT_SECRET", "409e31d3ddd64af08cfcc3b0f064fcbe"
)

# Only  Links formats are  accepted for this Var value.
SUPPORT_CHANNEL = getenv("SUPPORT_CHANNEL", None)  # Example:- https://t.me/TheTeamVivek
SUPPORT_GROUP = getenv("SUPPORT_GROUP", None)  # Example:- https://t.me/TheTeamVk


# Telegram audio  and video file size limit

TG_AUDIO_FILESIZE_LIMIT = int(
    getenv("TG_AUDIO_FILESIZE_LIMIT", "1073741824")
)  # Remember to give value in bytes

TG_VIDEO_FILESIZE_LIMIT = int(
    getenv("TG_VIDEO_FILESIZE_LIMIT", "1073741824")
)  # Remember to give value in bytes

# Chceckout https://www.gbmb.org/mb-to-bytes  for converting mb to bytes


# For customized or modified Repository
UPSTREAM_REPO = getenv(
    "UPSTREAM_REPO",
    "https://github.com/TheTeamVivek/YukkiMusic",
)
UPSTREAM_BRANCH = getenv("UPSTREAM_BRANCH", "master")

# Maximum number of video calls allowed on bot.
# You can later set it via /set_video_limit on telegram
VIDEO_STREAM_LIMIT = int(getenv("VIDEO_STREAM_LIMIT", "10"))

# Images

START_IMG_URL = getenv("START_IMG_URL", None)

PING_IMG_URL = getenv(
    "PING_IMG_URL",
    "assets/Ping.jpeg",
)

PLAYLIST_IMG_URL = getenv(
    "PLAYLIST_IMG_URL",
    "assets/Playlist.jpeg",
)

GLOBAL_IMG_URL = getenv(
    "GLOBAL_IMG_URL",
    "assets/Global.jpeg",
)

STATS_IMG_URL = getenv(
    "STATS_IMG_URL",
    "assets/Stats.jpeg",
)

TELEGRAM_AUDIO_URL = getenv(
    "TELEGRAM_AUDIO_URL",
    "assets/Audio.jpeg",
)

TELEGRAM_VIDEO_URL = getenv(
    "TELEGRAM_VIDEO_URL",
    "assets/Video.jpeg",
)

STREAM_IMG_URL = getenv(
    "STREAM_IMG_URL",
    "assets/Stream.jpeg",
)

SOUNCLOUD_IMG_URL = getenv(
    "SOUNCLOUD_IMG_URL",
    "assets/Soundcloud.jpeg",
)

YOUTUBE_IMG_URL = getenv(
    "YOUTUBE_IMG_URL",
    "assets/Youtube.jpeg",
)

SPOTIFY_ARTIST_IMG_URL = getenv(
    "SPOTIFY_ARTIST_IMG_URL",
    "assets/SpotifyArtist.jpeg",
)

SPOTIFY_ALBUM_IMG_URL = getenv(
    "SPOTIFY_ALBUM_IMG_URL",
    "assets/SpotifyAlbum.jpeg",
)

SPOTIFY_PLAYLIST_IMG_URL = getenv(
    "SPOTIFY_PLAYLIST_IMG_URL",
    "assets/SpotifyPlaylist.jpeg",
)


### DONT TOUCH or EDIT codes after this line

BANNED_USERS = filters.user()
YTDOWNLOADER = 1
LOG = 2
LOG_FILE_NAME = "logs.txt"
adminlist = {}
lyrical = {}
chatstats = {}
userstats = {}
clean = {}
autoclean = []


def time_to_seconds(time):
    stringt = str(time)
    return sum(int(x) * 60**i for i, x in enumerate(reversed(stringt.split(":"))))


def seconds_to_time(seconds):
    minutes = seconds // 60
    remaining_seconds = seconds % 60
    return f"{minutes:02d}:{remaining_seconds:02d}"


DURATION_LIMIT = int(time_to_seconds(f"{DURATION_LIMIT_MIN}:00"))
SONG_DOWNLOAD_DURATION_LIMIT = int(time_to_seconds(f"{SONG_DOWNLOAD_DURATION}:00"))

_DEFAULTS = {
    "PING_IMG_URL": "assets/Ping.jpeg",
    "PLAYLIST_IMG_URL": "assets/Playlist.jpeg",
    "GLOBAL_IMG_URL": "assets/Global.jpeg",
    "STATS_IMG_URL": "assets/Stats.jpeg",
    "TELEGRAM_AUDIO_URL": "assets/Audio.jpeg",
    "TELEGRAM_VIDEO_URL": "assets/Video.jpeg",
    "YOUTUBE_IMG_URL": "assets/Youtube.jpeg",
    "SOUNCLOUD_IMG_URL": "assets/Soundcloud.jpeg",
    "SPOTIFY_ALBUM_IMG_URL": "assets/SpotifyAlbum.jpeg",
    "SPOTIFY_ARTIST_IMG_URL": "assets/SpotifyArtist.jpeg",
    "SPOTIFY_PLAYLIST_IMG_URL": "assets/SpotifyPlaylist.jpeg",
    "STREAM_IMG_URL": "assets/Stream.jpeg",
}

_REQUIRED_URLS = [
    "EXTRA_PLUGINS_REPO",
    "SUPPORT_CHANNEL",
    "SUPPORT_GROUP",
    "UPSTREAM_REPO",
    "GITHUB_REPO",
]

for var_name, default_url in _DEFAULTS.items():
    url = globals().get(var_name)
    if url and url != default_url and not re.match(r"^https?://", url):
        print(
            f"[ERROR] - Your {var_name} URL is incorrect. Please ensure it starts with https://"
        )
        sys.exit()

for var_name in _REQUIRED_URLS:
    url = globals().get(var_name)
    if url and not re.match(r"^https?://", url):
        print(
            f"[ERROR] - Your {var_name} URL is incorrect. Please ensure it starts with https://"
        )
        sys.exit()
